File: 03-Sliding-Window/08-Repeated-DNA-Sequences/RepeatedDNASequences.java
/**
 * Repeated DNA Sequences
 * LeetCode: https://leetcode.com/problems/repeated-dna-sequences/
 * Time: O(n), Space: O(n)
 */

import java.util.*;

class RepeatedDNASequences {
    public List<String> findRepeatedDnaSequences(String s) {
        Set<String> seen = new HashSet<>();
        Set<String> repeated = new HashSet<>();

        for (int i = 0; i <= s.length() - 10; i++) {
            String sequence = s.substring(i, i + 10);
            if (seen.contains(sequence)) {
                repeated.add(sequence);
            }
            seen.add(sequence);
        }

        return new ArrayList<>(repeated);
    }

    public static void main(String[] args) {
        RepeatedDNASequences solution = new RepeatedDNASequences();
        System.out.println(solution.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"));
        // ["AAAAACCCCC","CCCCCAAAAA"]
    }
}

File: 03-Sliding-Window/08-Repeated-DNA-Sequences/RepeatedDNASequences.cpp
/**
 * Repeated DNA Sequences
 * Time: O(n), Space: O(n)
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class RepeatedDNASequences {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        unordered_set<string> seen, repeated;

        for (int i = 0; i <= (int)s.length() - 10; i++) {
            string sequence = s.substr(i, 10);
            if (seen.count(sequence)) {
                repeated.insert(sequence);
            }
            seen.insert(sequence);
        }

        return vector<string>(repeated.begin(), repeated.end());
    }
};

int main() {
    RepeatedDNASequences solution;
    auto result = solution.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT");
    for (const auto& seq : result) {
        cout << seq << " ";
    }
    cout << endl;
    return 0;
}
